File: 0141-linked-list-cycle/0141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    const set = new Set()
    let curr = head;
    while(curr) {
        if (set.has(curr)) return true;
        set.add(curr)
        curr = curr.next
    }
    return false
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        this_set = set()
        curr = head
        while (curr):
            if curr in this_set:
                return True
            this_set.add(curr)
            curr = curr.next
        return False